filters/auth/jwt_validation.go

```diff
@@ -0,0 +1,301 @@
+package auth
+
+import (
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-co-op/gocron"
+	"github.com/golang-jwt/jwt"
+	"github.com/opentracing/opentracing-go"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+)
+
+const (
+	JwtValidationName = "jwtValidation"
+)
+
+type (
+	jwtValidationSpec struct {
+		options TokenintrospectionOptions
+	}
+
+	jwtValidationFilter struct {
+		scheduler       *gocron.Scheduler
+		claims          []string
+		upstreamHeaders map[string]string
+	}
+)
+
+var m sync.Mutex
+var rsakeys map[string]*rsa.PublicKey = make(map[string]*rsa.PublicKey)
+
+func NewJwtValidationWithOptions(o TokenintrospectionOptions) filters.Spec {
+	return &jwtValidationSpec{
+		options: o,
+	}
+}
+
+func NewJwtValidation(timeout time.Duration) filters.Spec {
+	return NewJwtValidationWithOptions(TokenintrospectionOptions{
+		Timeout: timeout,
+		Tracer:  opentracing.NoopTracer{},
+	})
+}
+
+func JwtValidationWithOptions(
+	create func(time.Duration) filters.Spec,
+	o TokenintrospectionOptions,
+) filters.Spec {
+	s := create(o.Timeout)
+	ts, ok := s.(*jwtValidationSpec)
+	if !ok {
+		return s
+	}
+
+	ts.options = o
+	return ts
+}
+
+func (s *jwtValidationSpec) Name() string {
+	return JwtValidationName
+}
+
+func (s *jwtValidationSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) < 2 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	issuerURL := sargs[0]
+
+	cfg, err := getOpenIDConfig(issuerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	scheduler := gocron.NewScheduler(time.UTC)
+	scheduler.Every("30m").Do(func() {
+		err = fillPublicKeys(cfg.JwksURI)
+		if err != nil {
+			log.Errorf("Error getting public keys for url: %s", cfg.JwksURI)
+			return
+		}
+	})
+
+	scheduler.StartAsync()
+
+	f := &jwtValidationFilter{
+		scheduler: scheduler,
+	}
+
+	f.claims = strings.Split(sargs[1], " ")
+	if !all(f.claims, cfg.ClaimsSupported) {
+		return nil, fmt.Errorf("%v: %s, supported Claims: %v", errUnsupportedClaimSpecified, strings.Join(f.claims, ","), cfg.ClaimsSupported)
+	}
+
+	// inject additional headers from the access token for upstream applications
+	if len(sargs) > 2 && sargs[2] != "" {
+		f.upstreamHeaders = make(map[string]string)
+
+		for _, header := range strings.Split(sargs[2], " ") {
+			sl := strings.SplitN(header, ":", 2)
+			if len(sl) != 2 || sl[0] == "" || sl[1] == "" {
+				return nil, fmt.Errorf("%w: malformatted filter for upstream headers %s", filters.ErrInvalidFilterParameters, sl)
+			}
+			f.upstreamHeaders[sl[0]] = sl[1]
+		}
+		log.Debugf("Upstream Headers: %v", f.upstreamHeaders)
+	}
```

Comment:
I think this filter should do bare minimum - validate token and maybe certain claims.

Token (or its fields) forwarding is an optional usecase and should be implemented by separate filter. Maybe separate filter is not even needed and https://github.com/zalando/skipper/blob/master/docs/reference/filters.md#forwardtoken could be reused.

Reply:
I see your point. However in this case whole PR will be useless for me because I won't be able to cover my use-case.

Reply:
> do I understand properly that setContextRequestHeader filter is not working with maps

correct

> Maybe we can extend them to support map traversal

In such case `setContextRequestHeader` would have to know about statebag key name which would make it impossible to refactor later on. I would say statebag should be kept for internal communication within skipper, it is unfortunate `setContext*` are already there. See also related discussion https://github.com/zalando/skipper/pull/1528#discussion_r526505822 

I think we have the following options:
- use existing `forwardToken` and adjust backend to read json from the header value instead of plain string
- extend `forwardToken` somehow to produce non-json string response
- introduce separate filter to extract field in non-json format from the token payload

Reply:
Let's try to proceed with third option - create separate filter which extracts fields/claims from statebag with key "oidcclaimscachekey". Any suggestion on naming (set|forward|apply)(Token|HeaderFromToken)(Claim|Field) ?

Reply:
created a new filter forwardTokenPart and removed header functionality from jwtValidation